### PR TITLE
[SPARK-7058] Include RDD deserialization time in "task deserialization time" metric

### PR DESCRIPTION
**Pull request by JoshRosen**
The web UI's "task deserialization time" metric is slightly misleading because it does not capture the time taken to deserialize the broadcasted RDD.

### CODE REVIEW

**Comment by JoshRosen** [comment]
/cc @kayousterhout @rxin.

I noticed this in some benchmarking work that I'm doing (more details on the JIRA: https://issues.apache.org/jira/browse/SPARK-7058).  Before this patch, we would almost always report 1 or 2 millisecond deserialization times.  Now that this metric captures all of the deserialization costs, I'm seeing tasks that spend between 70 and 150ms in deserialization.

I have several ideas of how to optimize this deserialization to reduce this time and I'll address them later patches.

**Comment by JoshRosen** [comment]
As written here, I guess that this double-counts some of the time spent in execution, so I probably need to move the setting of the task start time into Task.  Let me make that change now.

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 1]
```diff
@@ -47,10 +47,9 @@ class TaskMetrics extends Serializable {
   /**
    * Time taken on the executor to deserialize this task
    */
-  private var _executorDeserializeTime: Long = _
+  private var _executorDeserializeTime: Long = 0
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: Why this change? Aren't these two things equivalent?

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: Yep, it's not necessary; will roll back to minimize diff.

**Comment by kayousterhout** [comment]
Thanks for fixing this @JoshRosen!  I've sometimes wondered if it would be helpful to specifically break out the broadcast time to help folks with debugging? In any case, this is a fine intermediate solution, even if we do decide to break out broadcast time eventually.

**Comment by JoshRosen** [comment]
I've updated this patch to push the calculation of the task run time into the Task itself; this avoids double-counting of the deserialization time, which was breaking the calculation of scheduler delay.

**Comment by kayousterhout** [comment]
It makes me a little nervous that there's now a time gap between deserializeEndTime and when taskStartTime gets calculated.  This *should* be very small (there's just the intermediate call to updateEpoch) but sometimes things like this get bigger over time (as code changes etc.), and that will make the metrics very confusing.  Can the task class expose executorDeserializeTime, and then Executor.scala can call that at the end to appropriately set all of the metrics?  I also slightly prefer that approach because it consolidates the metric setting to be mostly in Executor.scala.

**Comment by kayousterhout** [comment]
Also it is prohibitively difficult to write a unit test for this? I suspect the answer is yes...

**Comment by JoshRosen** [comment]
I think that the right way to unit test this would be to get the time via the `Clock` interface instead of calling `System.currentTimeMillis()` directly, create a static Clock instance somewhere, then rig objects to advance the time on this clock instance at various points, such as deserialization, task execution, etc.  This is totally doable, but it might involve a lot of changes to properly pass the Clock instance to the right places.

**Comment by JoshRosen** [comment]
Exposing the time from Task seems like a better design; I've updated to incorporate this idea.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 2]
```diff
@@ -221,8 +221,9 @@ private[spark] class Executor(
         val afterSerialization = System.currentTimeMillis()
 
         for (m <- task.metrics) {
-          m.setExecutorDeserializeTime(taskStart - deserializeStartTime)
-          m.setExecutorRunTime(taskFinish - taskStart)
+          m.setExecutorDeserializeTime(
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Nit: It would be super helpful to have a comment here to say something to describe the two components of this, like "Deserialization happens in two parts: first, we deserialize a Task object, which includes the Partition. Second, Task.run() deserializes the RDD and function to be run."

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala [hunk 3]
```diff
@@ -87,12 +87,19 @@ private[spark] abstract class Task[T](val stageId: Int, var partitionId: Int) ex
...
 
+  protected var _executorDeserializeTime: Long = 0
+
   /**
    * Whether the task has been killed.
    */
   def killed: Boolean = _killed
 
   /**
+   * Returns the amount of time spent on task / RDD deserialization.
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: maybe "spent deserializing the RDD and function to be run"? (since technically deserialization of the task happens in Executor.scala, not here)

**Comment by kayousterhout** [comment]
One more nit: could you update the task deserialization time tooltip to explicitly say that it includes the time to read the broadcasted task?

Other than that and the other two small comments I had, LGTM!

**Comment by kayousterhout** [comment]
LGTM!

**Comment by rxin** [comment]
LGTM!
